runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/BeamWorkerStatusGrpcService.java

```diff
@@ -92,90 +92,97 @@ public void close() throws Exception {
 
     WorkerStatusClient fnApiStatusClient =
         WorkerStatusClient.forRequestObserver(workerId, requestObserver);
-    fnApiStatusClient.setDeregisterCallback(this.connectedClient::remove);
-    if (connectedClient.containsKey(workerId) && connectedClient.get(workerId).isDone()) {
-      LOG.info(
-          "SDK Worker {} was connected to status server previously, disconnecting the old client",
-          workerId);
-      try {
-        WorkerStatusClient oldClient = connectedClient.get(workerId).get();
-        oldClient.close();
-      } catch (IOException | InterruptedException | ExecutionException e) {
-        LOG.warn("Error closing worker status client", e);
-      }
-    }
-    connectedClient
-        .computeIfAbsent(workerId, k -> new CompletableFuture<>())
-        .complete(fnApiStatusClient);
+    connectedClient.compute(
+        workerId,
+        (k, existingClientFuture) -> {
+          if (existingClientFuture != null) {
+            try {
+              if (existingClientFuture.isDone()) {
+                LOG.info(
+                    "SDK Worker {} was connected to status server previously, disconnecting old client",
+                    workerId);
+                existingClientFuture.get().close();
+              } else {
+                existingClientFuture.complete(fnApiStatusClient);
+                return existingClientFuture;
+              }
+            } catch (IOException | InterruptedException | ExecutionException e) {
+              LOG.warn("Error closing worker status client", e);
+            }
+          }
+          return CompletableFuture.completedFuture(fnApiStatusClient);
+        });
     return fnApiStatusClient.getResponseObserver();
   }
 
   /**
-   * Get the latest SDK worker status from the client's corresponding SDK Harness.
+   * Get the latest SDK worker status from the client's corresponding SDK harness.
    *
    * @param workerId worker id of the SDK harness.
    * @return {@link CompletableFuture} of WorkerStatusResponse from SDK harness.
    */
-  public CompletableFuture<WorkerStatusResponse> getWorkerStatus(String workerId) {
+  public String getSingleWorkerStatus(String workerId, long timeout, TimeUnit timeUnit) {
+    if (!connectedClient.containsKey(workerId)) {
```

Comment:
This check here is not needed since getWorkerStatus will return a completed future and getWorkerStatus needs to do that work otherwise you expose a race where connectedClient is mutated after this check but before getWorkerStatus is invoked.

Reply:
done.